Add nilu air_quality platform

hfurubotten opened the pull request:
## Description:

New platform in the air_quality platform. Shows data from the governmental air research institute sensor stations within Norway.

Rewritten from a `sensor` platfrom to `air_quality` platform. Original PR here: #18104

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7993

## Example entry for `configuration.yaml` (if applicable):
```yaml
air_quality:
  - platform: nilu
    show_on_map: True
  - platform: nilu
    show_on_map: True
    area: Bergen
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

hfurubotten committed:
Add nilu air_pollutants platform

MartinHjelmare commented on homeassistant/components/air_pollutants/nilu.py:
```diff
@@ -0,0 +1,311 @@
+    def carbon_monoxide(self) -> str:
+        """Return the CO (carbon monoxide) level."""
+        from niluclient import CO
+        if CO in self._api.data.sensors:
```
Extract the rest of the property code into a helper method that accepts the measurement constant and returns the value or `None`. Almost all the properties use the same pattern.

MartinHjelmare commented on homeassistant/components/air_pollutants/nilu.py:
```diff
@@ -0,0 +1,311 @@
+        return self._name
+
+    @property
+    def state(self) -> str:
```
Don't overwrite this. The base entity class will handle this.

MartinHjelmare commented on homeassistant/components/air_pollutants/nilu.py:
```diff
@@ -0,0 +1,311 @@
+import voluptuous as vol
+
+from homeassistant.components.air_pollutants import AirPollutantsEntity
+from homeassistant.components.sensor import PLATFORM_SCHEMA
```
Import tge platform schema from the air pollutants component.

MartinHjelmare commented on homeassistant/components/air_pollutants/nilu.py:
```diff
@@ -0,0 +1,311 @@
+    vol.Exclusive(CONF_STATION, 'station_collection',
+                  'Can only configure one specific station or '
+                  'stations in a specific area pr sensor. '
+                  'Please only configure station or area.'): cv.ensure_list,
```
Validate the item in the list too.

hfurubotten committed:
Code Review - validation, DRYs, rm state override, new attr
- Repeated code moved to own method.
- Removed override of state property.
- New attr for showing nilu pollution recommendations.
- More validation of stations input.
- Minor fixes and typos.

hfurubotten commented on homeassistant/components/air_pollutants/nilu.py:
```diff
@@ -0,0 +1,265 @@
+        """Get data cached in client."""
+        return self.api.data
+
+    @Throttle(SCAN_INTERVAL)
```
Sort of, yes. The source only aggregates and publishes new data once an hour, so no need to request the API too often unnecessary. There is no rate limit, but no need to overload them 🙂 A couple of times an hour I believe would be enough to get good refresh time and at the same time protect the source a little bit.

hfurubotten replied:
btw, is it possible to tell the update to work with an cron instead of an time interval? Then a cron with `5 */1 * * *` would be much better to achieve this.

MartinHjelmare replied:
If we want a different interval than a set time interval, we can take control and schedule the interval in the platform.

I don't know by heart what that cron interval means.

hfurubotten replied:
I'm still a bit green on all the inner workings of HA, so don't know how to do the manual scheduling during setup, but the idea was to do the update with "minute 5 past every hour" as most of the stations seams to publish around the hour, or +/- 5 min.  
But every 30 or 15 min would not mean too much extra requests, and considering extra complexity to scheduling, it might not be worth it?

MartinHjelmare replied:
Yeah, I suggest setting it to 1 hour interval.

hfurubotten committed:
Removed unused prop

MartinHjelmare commented on homeassistant/components/air_pollutants/nilu.py:
```diff
@@ -0,0 +1,265 @@
+        stations = location_client.station_names
+
+    for station in stations:
+        client = NiluData(nilu.create_station_client(station))
```
Yes, we shouldn't create an entity that is useless. Even better would be to catch this during config validation, but maybe that isn't possible?

MartinHjelmare replied:
Yes, a warning is good and then not add that station/entity.

hfurubotten replied:
I will fix it when I get a free moment tomorrow. Thanks and Happy New year! 🎉

hfurubotten committed:
Check for none result from client before entity add

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Let's wait here until we know if https://github.com/home-assistant/home-assistant/pull/19448 should be merged first.

MartinHjelmare commented:
#19448 has been merged so we need to rebase and move this platform under the `air_quality` component.

hfurubotten committed:
Moved platform to air_quality component

hfurubotten commented:
It should be rebased and moved to the air_quality component now. 
Please check that everything is as it should. Got myself into trouble when rebasing, but I think I have gotten everything into order. However a double check would be nice.

MartinHjelmare commented on homeassistant/components/air_quality/nilu.py:
```diff
@@ -0,0 +1,268 @@
+Data delivered by luftkvalitet.info and nilu.no.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/air_pollutants.nilu/
```
Stale url.

MartinHjelmare commented on homeassistant/components/air_quality/nilu.py:
```diff
@@ -0,0 +1,268 @@
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the NILU air pollutants sensor."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/air_quality/nilu.py:
```diff
@@ -0,0 +1,268 @@
+        if component_name in self._api.data.sensors:
+            sensor = self._api.data.sensors[component_name]
+            return "{0:.2f} {1}".format(sensor.value,
+                                        sensor.unit_of_measurement)
```
There should preferably not be a unit of measurement but that should be standardized. Maybe we haven't done that yet? @fabaff

hfurubotten replied:
Is it possible to add en property where it gives the unit of measurement, such as on sensors? I was not sure if all measured components should be defined in `µg/m³`, so I added it to the output of each.  

All measured pollutants currently supported from the NILU-client is measured in `µg/m³` or `mg/m³`, so one fit it all is possible with what the `air_quality` component gives of information now. `CO`/carbon monoxide is the one thats measured in `mg/m³`, but this can easily be converted to `µg/m³` to get a one fits it all.

Can add this to the platform:
```python
    @property
    def unit_of_measurement(self) -> str:
        """Return the unit this state is expressed in."""
        return 'µg/m³'
```

fabaff replied:
There should be no unit of measurements because the frontend should take care about this.

hfurubotten replied:
I will fix the output. What is the default unit the frontend expect, so I can make sure they are correct or converted?

hfurubotten committed:
Updated outdated docstrings

fabaff committed:
Minor changes
